feat(curriculum): add Sudoku solver project Python (#52452)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/20-upcoming-python/learn-classes-and-objects-by-building-a-sudoku-solver/6569fd6d3cb95856c9ed2190.md

---
id: 6569fd6d3cb95856c9ed2190
title: Step 77
challengeType: 20
dashedName: step-77
---

# --description--

> If not `True`, create an `else` statement to print a message that says "The provided puzzle is unsolvable."

Reviewer: ```suggestion
Create an `else` clause and print the following string inside the new `else` block: `'\nThe provided puzzle is unsolvable.'`.
```

# --hints--

> Create an `else` statement to print a message that says "The provided puzzle is unsolvable."

Reviewer: ```suggestion
You should create an `else` clause and print the provided string inside the `else` block.
```

```js

> const after_split = code.split('if gameboard.solver():')[1];
> assert.match(after_split, /else/);

Reviewer: ```suggestion
({ test: () => assert.match(e.code.original["main.py"], /^(\s{4})else\s*:\s*^\1\1print\s*\(\s*("|')\\nThe\sprovided\spuzzle\sis\sunsolvable\.\2\s*\)/m) })
```

```

> 
> Print the message "The provided puzzle is unsolvable." using `print("The provided puzzle is unsolvable.")`.
> 
> ```js
> assert.match(code, /print\("The provided puzzle is unsolvable."\)/);
> ```

Reviewer: ```suggestion
```
removed as a consequence of the previous comment

        return all(
            self.board[row][col] != num
            for row in range(9)
        )

    def valid_in_square(self, row, col, num):
        row_start = (row // 3) * 3
        col_start=(col // 3) * 3
        for row_no in range(row_start, row_start + 3):
            for col_no in range(col_start, col_start + 3):
                if self.board[row_no][col_no] == num:
                    return False
        return True

    def is_valid(self, empty, num):
        row, col = empty
        valid_in_row = self.valid_in_row(row, num)
        valid_in_col = self.valid_in_col(col, num)
        valid_in_square = self.valid_in_square(row, col, num)
        return all([valid_in_row, valid_in_col, valid_in_square])

    def solver(self):
        if (next_empty := self.find_empty_cell()) is None:
            return True
        else:
            for guess in range(1, 10):
                if self.is_valid(next_empty, guess):
                    row, col = next_empty
                    self.board[row][col] = guess
                    if self.solver():
                        return True
                    self.board[row][col] = 0

        return False

> def solve_sudoku(board):
>     gameboard = Board(board)
>     print('\nPuzzle to solve:')
>     print(gameboard)
>     if gameboard.solver():
>         print('\nSolved puzzle:')
>         print(gameboard)
> --fcc-editable-region--

Reviewer: ```suggestion
--fcc-editable-region--
def solve_sudoku(board):
    gameboard = Board(board)
    print('\nPuzzle to solve:')
    print(gameboard)
    if gameboard.solver():
        print('\nSolved puzzle:')
        print(gameboard)
```
